embassy-stm32/src/flash/mod.rs

```diff
@@ -0,0 +1,401 @@
+    pub(crate) const FLASH_START: usize = FLASH_BASE;
+    pub(crate) const FLASH_END: usize = FLASH_START + FLASH_SIZE;
+    pub(crate) const PAGE_SIZE: usize = 2048;
+    pub(crate) const WORD_SIZE: usize = 8;
+}
+
+#[cfg(flash_wb)]
+mod config {
+    use super::*;
+    pub(crate) const FLASH_SIZE: usize = 0x100000;
+    pub(crate) const FLASH_START: usize = FLASH_BASE;
+    pub(crate) const FLASH_END: usize = FLASH_START + FLASH_SIZE;
+    pub(crate) const PAGE_SIZE: usize = 4096;
+    pub(crate) const WORD_SIZE: usize = 8;
+}
+
+#[cfg(flash_wl)]
+mod config {
+    use super::*;
+    pub(crate) const FLASH_SIZE: usize = 0x40000;
+    pub(crate) const FLASH_START: usize = FLASH_BASE;
+    pub(crate) const FLASH_END: usize = FLASH_START + FLASH_SIZE;
+    pub(crate) const PAGE_SIZE: usize = 2048;
+    pub(crate) const WORD_SIZE: usize = 8;
+}
+
+#[cfg(flash_l0)]
+mod config {
+    use super::*;
+    pub(crate) const FLASH_SIZE: usize = 0x30000;
+    pub(crate) const FLASH_START: usize = FLASH_BASE;
+    pub(crate) const FLASH_END: usize = FLASH_START + FLASH_SIZE;
+    pub(crate) const PAGE_SIZE: usize = 128;
+    pub(crate) const WORD_SIZE: usize = 4;
+}
+
+#[cfg(flash_l1)]
+mod config {
+    use super::*;
+    pub(crate) const FLASH_SIZE: usize = 0x80000;
+    pub(crate) const FLASH_START: usize = FLASH_BASE;
+    pub(crate) const FLASH_END: usize = FLASH_START + FLASH_SIZE;
+    pub(crate) const PAGE_SIZE: usize = 256;
+    pub(crate) const WORD_SIZE: usize = 4;
+}
+
+use config::*;
+
+pub struct Flash<'d> {
+    _inner: FLASH,
+    _phantom: PhantomData<&'d mut FLASH>,
+}
+
+impl<'d> Flash<'d> {
+    pub fn new(p: impl Unborrow<Target = FLASH>) -> Self {
+        unborrow!(p);
+        Self {
+            _inner: p,
+            _phantom: PhantomData,
+        }
+    }
+
+    pub fn unlock(p: impl Unborrow<Target = FLASH>) -> Self {
+        let flash = Self::new(p);
+        #[cfg(any(flash_wl, flash_wb, flash_l4))]
+        unsafe {
+            pac::FLASH.keyr().write(|w| w.set_keyr(0x4567_0123));
+            pac::FLASH.keyr().write(|w| w.set_keyr(0xCDEF_89AB));
+        }
+
+        #[cfg(any(flash_l0))]
+        unsafe {
+            pac::FLASH.pekeyr().write(|w| w.set_pekeyr(0x89ABCDEF));
+            pac::FLASH.pekeyr().write(|w| w.set_pekeyr(0x02030405));
+
+            pac::FLASH.prgkeyr().write(|w| w.set_prgkeyr(0x8C9DAEBF));
+            pac::FLASH.prgkeyr().write(|w| w.set_prgkeyr(0x13141516));
+        }
+        flash
+    }
+
+    pub fn lock(&mut self) {
+        #[cfg(any(flash_wl, flash_wb, flash_l4))]
+        unsafe {
+            pac::FLASH.cr().modify(|w| w.set_lock(true));
+        }
+
+        #[cfg(any(flash_l0))]
+        unsafe {
+            pac::FLASH.pecr().modify(|w| {
+                w.set_optlock(true);
+                w.set_prglock(true);
+                w.set_pelock(true);
+            });
+        }
+    }
+
+    pub fn blocking_read(&mut self, offset: u32, bytes: &mut [u8]) -> Result<(), Error> {
```

Comment:
the embedded-storage trait specifies `offset` is relative to the start of the memory. Even if the flash is memory-mapped starting at `0x8000000`, the offset should start at `0`.